### PR TITLE
Migrate more options to VerifierBuilder

### PR DESCRIPTION
**Pull request by pavel-mikula-sonarsource**
Contributes to #5186 

Add more options:
*  `.WithErrorBehavior`
*  `.WithOnlyDiagnostics`
*  `.WithOutputKind`
*  `.WithTopLevelSupport`
*  `.WithSonarProjectConfig`

This is a draft, rebase will be needed after #5208 is merged. The 1st squash commit doesn't need a review for the same reason.

### CODE REVIEW

**Comment by Corniel** [comment]
Don't forget to add `.AddSnippet()`.

Also, C# and VB.NET have some features that do not apply to other languages (like allow unsafe code). It might be worth to have language specific builders (that derive from the same base), as I did [here](https://github.com/Corniel/roslyn-test-tool). I like the approach, including being immutable.

**Comment by pavel-mikula-sonarsource** [comment]
@Corniel The current `OldVerifier` is not fully replaced yet, more PRs will come.

C# vs VB is currently solved in SolutionBuilder and doesn't pass through Verifier.

**Comment by Corniel** [comment]
> C# vs VB is currently solved in SolutionBuilder and doesn't pass through Verifier.

THe difference should (indeed) not be in the verifier, but in the builder, that creates the verifier.

File: analyzers/tests/SonarAnalyzer.UnitTest/TestFramework/OldVerifier.cs [hunk 1]
```diff
@@ -20,6 +20,7 @@
 
 using System;
 using System.Collections.Generic;
+using System.Collections.Immutable;
 using System.IO;
 using System.Linq;
 using System.Text;
```
Comment (costin-zaharia-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/TestFramework/OldVerifier.cs:26]: `using SonarAnalyzer.UnitTest.Helpers;` is no longer needed

File: analyzers/tests/SonarAnalyzer.UnitTest/TestFramework/OldVerifier.cs [hunk 2]
```diff
@@ -301,68 +302,104 @@ public static void VerifyNoExceptionThrown(string path, DiagnosticAnalyzer[] dia
...
         public static void VerifyAnalyzer(string path,
                                           DiagnosticAnalyzer[] diagnosticAnalyzers,
-                                          IEnumerable<ParseOptions> options = null,
+                                          ImmutableArray<ParseOptions> options = default,
                                           CompilationErrorBehavior checkMode = CompilationErrorBehavior.Default,
                                           OutputKind outputKind = OutputKind.DynamicallyLinkedLibrary,
                                           IEnumerable<MetadataReference> additionalReferences = null,
                                           DiagnosticDescriptor[] onlyDiagnostics = null) =>
-            VerifyAnalyzer(new[] { path }, diagnosticAnalyzers, options, checkMode, outputKind, additionalReferences, null, onlyDiagnostics);
+            diagnosticAnalyzers.Aggregate(new VerifierBuilder(), (builder, analyzer) => builder.AddAnalyzer(() => analyzer))
```
Comment (costin-zaharia-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/TestFramework/OldVerifier.cs:313]: That's a lot of builders created without a real need. I think we should either change `AddAnalyzer` to be `AddAnalyzers` and accept more analyzers, or add an overload.

File: analyzers/tests/SonarAnalyzer.UnitTest/TestFramework/OldVerifier.cs [hunk 3]
```diff
@@ -301,68 +302,104 @@ public static void VerifyNoExceptionThrown(string path, DiagnosticAnalyzer[] dia
...
+                .AddReferences(additionalReferences)
+                .AddPaths(RemoveTestCasesPrefix(path))
+                .WithOptions(options.IsDefault ? ImmutableArray<ParseOptions>.Empty : options)
+                .WithErrorBehavior(checkMode)
+                .WithOutputKind(outputKind)
+                .WithOnlyDiagnostics(onlyDiagnostics ?? Array.Empty<DiagnosticDescriptor>())
+                .Verify();
```
Comment (costin-zaharia-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/TestFramework/OldVerifier.cs:320]: pedantic: The alignment is a bit off

File: analyzers/tests/SonarAnalyzer.UnitTest/TestFramework/OldVerifier.cs [hunk 4]
```diff
@@ -301,68 +302,104 @@ public static void VerifyNoExceptionThrown(string path, DiagnosticAnalyzer[] dia
...
-                           null, CompilationErrorBehavior.Default,
-                           OutputKind.DynamicallyLinkedLibrary,
-                           additionalReferences,
-                           sonarProjectConfigPath);
+            new VerifierBuilder()
+                .AddAnalyzer(() => diagnosticAnalyzer)
+                .AddReferences(additionalReferences)
+                .AddPaths(RemoveTestCasesPrefix(path))
+                .WithSonarProjectConfigPath(sonarProjectConfigPath)
+                .Verify();
```
Comment (costin-zaharia-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/TestFramework/OldVerifier.cs:332]: pedantic: We usually make the first method invocation on the same line and then align the rest of the calls with the first one.

File: analyzers/tests/SonarAnalyzer.UnitTest/TestFramework/Tests/IssueLocationCollectorTest.GetPreciseIssueLocations.cs [hunk 5]
```diff
@@ -19,18 +19,14 @@
...
 using Microsoft.VisualStudio.TestTools.UnitTesting;
 
-namespace SonarAnalyzer.UnitTest.TestFramework.IssueLocationCollectorTests
+namespace SonarAnalyzer.UnitTest.TestFramework.Tests
 {
-    [TestClass]
-    public class IssueLocationCollector_GetPreciseIssueLocations
+    public partial class IssueLocationCollectorTest
     {
-        [TestMethod]
```
Comment (costin-zaharia-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/TestFramework/Tests/IssueLocationCollectorTest.GetPreciseIssueLocations.cs:33]: Did you remove this by mistake? If I put it back, the test is passing.

File: analyzers/tests/SonarAnalyzer.UnitTest/TestFramework/Tests/VerifierTest.cs [hunk 6]
```diff
@@ -144,10 +147,57 @@ public void Verify_ParseOptions()
                 .WithMessage("CSharp5: Unexpected build error [CS8026]: Feature 'target-typed object creation' is not available in C# 5. Please use language version 9.0 or greater. on line 3");
         }
 
+        [TestMethod]
+        public void Verify_ErrorBehavior()
+        {
+            var builder = WithSnippet("undefined");
+            builder.Invoking(x => x.Verify()).Should().Throw<UnexpectedDiagnosticException>()
+                .WithMessage("CSharp7: Unexpected build error [CS0116]: A namespace cannot directly contain members such as fields, methods or statements on line 1");
```
Comment (costin-zaharia-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/TestFramework/Tests/VerifierTest.cs:155]: pedantic: alignment if off

File: analyzers/tests/SonarAnalyzer.UnitTest/TestFramework/VerifierBuilder.cs [hunk 7]
```diff
@@ -58,11 +68,34 @@ private VerifierBuilder(VerifierBuilder original)
             new(this) { Paths = Paths.Concat(paths).ToImmutableArray() };
 
         public VerifierBuilder AddReferences(IEnumerable<MetadataReference> references) =>
-            new(this) { References = References.Concat(references).ToImmutableArray() };
+            references != null && references.Any()
```
Comment (costin-zaharia-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/TestFramework/VerifierBuilder.cs:71]: we evaluate `references` enumerable multiple times.

File: analyzers/tests/SonarAnalyzer.UnitTest/TestFramework/VerifierBuilder.cs [hunk 8]
```diff
@@ -58,11 +68,34 @@ private VerifierBuilder(VerifierBuilder original)
...
         public VerifierBuilder WithOptions(ImmutableArray<ParseOptions> parseOptions) =>
             new(this) { ParseOptions = parseOptions };
 
+        public VerifierBuilder WithOutputKind(OutputKind outputKind) =>
+            new(this) { OutputKind = outputKind };
+
+        public VerifierBuilder WithSonarProjectConfigPath(string sonarProjectConfigPath) =>
+            new(this) { SonarProjectConfigPath = sonarProjectConfigPath };
+
+        public VerifierBuilder WithTopLevelStatements() =>
```
Comment (costin-zaharia-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/TestFramework/VerifierBuilder.cs:96]: I would remove this method:
- it's misleading (it should be ensure that the language versions contains C# 9 and no other previous version)
- can be called also for VB
- we can use directly `WithOutputKind(OutputKind.ConsoleApplication)` which is more clear about the behavior

File: analyzers/tests/SonarAnalyzer.UnitTest/ResourceTests/GeneratedResourcesTest.cs [hunk 9]
```diff
@@ -1,124 +0,0 @@
...
-using Microsoft.VisualStudio.TestTools.UnitTesting;
-using SonarAnalyzer.Common;
-using SonarAnalyzer.Helpers;
-using SonarAnalyzer.Rules;
-using SonarAnalyzer.Utilities;
-
-namespace SonarAnalyzer.UnitTest.ResourceTests
-{
-    [TestClass]
-    public class GeneratedResourcesTest
```
Comment (costin-zaharia-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/ResourceTests/GeneratedResourcesTest.cs:37]: Not sure I understand why this was removed. It does not seem related to this PR.

**Review by costin-zaharia-sonarsource — CHANGES_REQUESTED** [review]

File: analyzers/tests/SonarAnalyzer.UnitTest/SonarAnalyzer.UnitTest.csproj [hunk 10]
```diff
@@ -67,25 +67,6 @@
     <None Remove="CBDEMocks\CBDEWaitAndSucceeds.exe" />
   </ItemGroup>
 
-  <ItemGroup>
```
[the remark this answers is not in the document]
Comment (pavel-mikula-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/SonarAnalyzer.UnitTest.csproj:70]: _Short:_ It's not needed. 

_Middle:_ We don't need to use those resources in our UT project. Those autogenerated files are only meant to be TestCases.

_TL;DR;_ 
Check the [last commit](https://github.com/SonarSource/sonar-dotnet/pull/5213/commits/c3ba8ce0fd8e08e924f40da7fa2761a7de5dd396) in this PR above to see the isolated change.

`MarkAssemblyWithNeutralResourcesLanguageAttributeTest` used resx designer files as part of the UT. And while it only needed `just another cs file` that should be in `TestCases` directory, it was using `resx` file and its autogenerated designer file in `Resources` directory.

And as all test cases should be in `TestCases` directory, this irregularity poped up and revealed itself.

File: analyzers/tests/SonarAnalyzer.UnitTest/TestFramework/OldVerifier.cs (hunk 2 above)
Comment (pavel-mikula-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/TestFramework/OldVerifier.cs:313]: This `Aggregate` hack should only exist in the `OldVerifier` so I would not add a special overload for it. And I plan to remove these in the 1st round of cleanup - there's a bunch of overloads used only a few times that I'd like to remove in another PR.
This specific one is used only once with 2 analyzers. So it's not very expensive.

File: analyzers/tests/SonarAnalyzer.UnitTest/TestFramework/OldVerifier.cs (hunk 3 above)
Comment (pavel-mikula-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/TestFramework/OldVerifier.cs:320]: I'm probably not looking at the correct place and I don't see it. What is off exactly?

File: analyzers/tests/SonarAnalyzer.UnitTest/TestFramework/OldVerifier.cs (hunk 4 above)
Comment (pavel-mikula-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/TestFramework/OldVerifier.cs:332]: While I usually prefer less lines, I find this alignment easier to understand in this case. All modifications start on the same indent here.

File: analyzers/tests/SonarAnalyzer.UnitTest/TestFramework/Tests/IssueLocationCollectorTest.GetPreciseIssueLocations.cs (hunk 5 above)
Comment (pavel-mikula-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/TestFramework/Tests/IssueLocationCollectorTest.GetPreciseIssueLocations.cs:33]: oops :)

File: analyzers/tests/SonarAnalyzer.UnitTest/TestFramework/Tests/VerifierTest.cs (hunk 6 above)
Comment (pavel-mikula-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/TestFramework/Tests/VerifierTest.cs:155]: It is meant to be like this according to the `How we work` from 2021-12-22. It's not documented yet, I have an action item to document it.

File: analyzers/tests/SonarAnalyzer.UnitTest/TestFramework/OldVerifier.cs (hunk 4 above)
Comment (pavel-mikula-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/TestFramework/OldVerifier.cs:332]: I now better see what you meant based on this comment:
https://github.com/SonarSource/sonar-dotnet/pull/5213#discussion_r774479132

File: analyzers/tests/SonarAnalyzer.UnitTest/TestFramework/OldVerifier.cs (hunk 3 above)
Comment (pavel-mikula-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/TestFramework/OldVerifier.cs:320]: I now better see what you meant based on this comment:
https://github.com/SonarSource/sonar-dotnet/pull/5213#discussion_r774479132

File: analyzers/tests/SonarAnalyzer.UnitTest/TestFramework/VerifierBuilder.cs (hunk 7 above)
Comment (pavel-mikula-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/TestFramework/VerifierBuilder.cs:71]: I gave it a try, but it's not very nice. Should I just remove the `Any` check instead?

File: analyzers/tests/SonarAnalyzer.UnitTest/TestFramework/VerifierBuilder.cs (hunk 8 above)
Comment (pavel-mikula-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/TestFramework/VerifierBuilder.cs:96]: The next PR #5214 adds `.WithConcurrentAnalysis(false);` that needs to be always called with this.
I'll improve the C# 9 part.
I'm fine with ignoring the VB design flaw :)

File: analyzers/tests/SonarAnalyzer.UnitTest/ResourceTests/GeneratedResourcesTest.cs (hunk 9 above)
Comment (pavel-mikula-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/ResourceTests/GeneratedResourcesTest.cs:37]: Doesn't seem intentional at all.

Comment (pavel-mikula-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/ResourceTests/GeneratedResourcesTest.cs:37]: Ok, so I probably deleted it by accident when moving files out of `ResourcesTest` directory. 
The `ResourcesTest` directory is meant to be a data-only directory, see this
https://github.com/SonarSource/sonar-dotnet/blob/0cd6231a8756eaee401ba9dfcc5b7f4d37165e9d/analyzers/tests/SonarAnalyzer.UnitTest/SonarAnalyzer.UnitTest.csproj#L102-L104

So we should rename it to `TestResources` instead (to have it near `TestCases`) in another PR.

I've moved the file to another directory to decofuse its location.

File: analyzers/tests/SonarAnalyzer.UnitTest/TestFramework/Tests/VerifierTest.cs (hunk 6 above)
Comment (costin-zaharia-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/TestFramework/Tests/VerifierTest.cs:155]: My understanding was that we discussed "method declaration arguments" and not about this. I find aligned indentations a lot easier to read.

File: analyzers/tests/SonarAnalyzer.UnitTest/TestFramework/VerifierBuilder.cs (hunk 7 above)
Comment (costin-zaharia-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/TestFramework/VerifierBuilder.cs]: Yes. Does it bring any value :) (pun intended)

File: analyzers/tests/SonarAnalyzer.UnitTest/TestFramework/VerifierBuilder.cs (hunk 8 above)
Comment (costin-zaharia-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/TestFramework/VerifierBuilder.cs]: I'm still not sure about the value of this method. We can easily live without it

**Review by costin-zaharia-sonarsource — COMMENTED** [review]
Waiting for the removal of `Any` call and let's discuss directly the alignment and the `WithTopLevelStatements` method.

File: analyzers/tests/SonarAnalyzer.UnitTest/TestFramework/Tests/VerifierTest.cs (hunk 6 above)
Comment (pavel-mikula-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/TestFramework/Tests/VerifierTest.cs:155]: We've discussed both, see notes.
_Method declaration:_ We've decided not to change anything. Mary will investigate VS tools.
_Identation on chained invocation calls:_ This is the one that applies here and that I have the action item for to document it.

File: analyzers/tests/SonarAnalyzer.UnitTest/TestFramework/VerifierBuilder.cs (hunk 8 above)
Comment (pavel-mikula-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/TestFramework/VerifierBuilder.cs]: We could live without it, and it will help us many times to save complicated copy/paste scaffolding that will always go together (considering the #5214 modification)

File: analyzers/tests/SonarAnalyzer.UnitTest/TestFramework/VerifierBuilder.cs (hunk 7 above)
Comment (pavel-mikula-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/TestFramework/VerifierBuilder.cs]: We can actually simplify it even more, as the null could come only from the `OldVerifier` anyway and should not happen under normal usage.

File: analyzers/tests/SonarAnalyzer.UnitTest/TestFramework/OldVerifier.cs [hunk 11]
```diff
@@ -485,7 +485,7 @@ static IEnumerable<TMessage> ReadProtobuf(string path)
         {
             var builder = SolutionBuilder.Create()
                 .AddProject(AnalyzerLanguage.FromPath(path), outputKind: outputKind)
-                .AddReferences(additionalReferences)
+                .AddReferences(additionalReferences ?? Enumerable.Empty<MetadataReference>())
```
Comment (costin-zaharia-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/TestFramework/OldVerifier.cs:488]: I would add this directly in the method to avoid duplication in the calls.

File: analyzers/tests/SonarAnalyzer.UnitTest/TestFramework/Tests/VerifierBuilderTest.cs [hunk 12]
```diff
@@ -74,14 +74,6 @@ public void AddReferences_Concatenates_IsImmutable()
             two.References.Should().BeEquivalentTo(MetadataReferenceFacade.MsCorLib.Concat(MetadataReferenceFacade.SystemData)).And.HaveCount(2);
         }
 
-        [TestMethod]
```
Comment (costin-zaharia-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/TestFramework/Tests/VerifierBuilderTest.cs:72]: And then we could leave these.

File: analyzers/tests/SonarAnalyzer.UnitTest/TestFramework/VerifierBuilder.cs [hunk 13]
```diff
@@ -68,9 +68,7 @@ private VerifierBuilder(VerifierBuilder original)
             new(this) { Paths = Paths.Concat(paths).ToImmutableArray() };
 
         public VerifierBuilder AddReferences(IEnumerable<MetadataReference> references) =>
-            references?.ToArray() is { Length: not 0 } referenceArray
-                ? new(this) { References = References.Concat(referenceArray).ToImmutableArray() }
-                : this;
+            new(this) { References = References.Concat(references).ToImmutableArray() };
```
Comment (costin-zaharia-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/TestFramework/VerifierBuilder.cs:71]: nice, much shorter.

**Review by costin-zaharia-sonarsource — APPROVED** [review]
LGTM! Would be great to reduce the duplication (see my comments) before merging.

File: analyzers/tests/SonarAnalyzer.UnitTest/TestFramework/OldVerifier.cs (hunk 11 above)
Comment (pavel-mikula-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/TestFramework/OldVerifier.cs:488]: It's duplicate and `OldVerifier` is supposed to go away. I'd like to keep `OldVerifier`-specific pain only there to keep the `VerifierBuilder` clean.
